src/lib.rs

```diff
@@ -22,35 +23,39 @@ struct Bloom {
     /// Equivalent to the number of hash function in the classic bloom filter.
     /// An insertion will result in a bit being set in each slice.
     #[serde(rename = "k")]
-    num_slices: usize,
+    num_slices: NonZeroU64,
 }
 
 impl Bloom {
-    /// Create a new Bloom filter
+    /// Create a new Bloom filter (specifically, a Partitioned Bloom filter)
     ///
     /// # Arguments
     ///
     /// * `capacity` - target capacity.
-    /// * `error_ratio` - false positive ratio [0..1.0].
+    /// * `error_ratio` - false positive ratio (0..1.0).
     /// * `seed` - a seed to be used to initialize the hasher.
     fn new(capacity: usize, error_ratio: f64) -> Bloom {
+        // Directly from paper:
+        // k = log2(1/P)   (num_slices)
+        // n ≈ −m ln(1−p)  (slice_len_bits)
```

Comment:
I'm a little spooked about this approximation, as the paper says 'For usable values of _m_` - are our slice lengths really that large? The paper says: "For any given error probability _P_ and filter size _M_, _n_ is maximized by making _p= 1/2_, regardless of _P_ or _M_."

I think the old formula is better, and I vaguely recall other libraries using the old formula

Reply:
This is tricky. It's important to consider that all formulas are approximations. The problem with the previous approach/formula is that it's derived from the optimal value of K, which is unrealistic since K _must_ be an integer. The end result is that the resulting filter would have a slightly off false-positive ratio.

The way I have it now will at least guarantee we get a BF with a false-positive rate <= the desired one, which is important to get a correct "upper-bound" when growing the scalable bloom filter.

Bloom filters are incredibly easy to get "slightly wrong" honestly. I'm doing my best to get this one right.

Reply:
Sounds good, and that's fair. I'm pretty comfortable with the changes at this stage. I'm not sure if you're still experimenting/making changes, so lemme know when you're comfortable with the changes

Reply:
I think it's good to go, lets leave it open for a few days just in case.

Reply:
I just realized that we also need something like this https://github.com/rust-lang/rust/blob/c0955a34bcb17f0b31d7b86522a520ebe7fa93ac/src/librustc_data_structures/stable_hasher.rs#L78-L166 to wrap the hasher. Otherwise hashing integers will have different outputs in LE/BE systems.